src/main/java/com/nutomic/syncthingandroid/activities/ShareActivity.java

```diff
@@ -0,0 +1,283 @@
+
+        if (!ContentResolver.SCHEME_CONTENT.equals(uri.getScheme())) {
+            displayName = uri.getLastPathSegment();
+        } else {
+            displayName = getDisplayNameFromContentResolver(uri);
+            try {
+                if (displayName == null) {
+                    // last chance to have a name
+                    displayName = uri.getLastPathSegment().replaceAll("\\s", "");
+                }
+
+                // Add best possible extension
+                int index = displayName.lastIndexOf(".");
+                if (index == -1 || MimeTypeMap.getSingleton()
+                        .getMimeTypeFromExtension(displayName.substring(index + 1)) == null) {
+                    String mimeType = this.getContentResolver().getType(uri);
+                    String extension = MimeTypeMap.getSingleton()
+                            .getExtensionFromMimeType(mimeType);
+                    if (extension != null) {
+                        displayName += "." + extension;
+                    }
+                }
+            } catch (Exception e) {
+                Log.e(TAG, "No way to get a display name for " + uri.toString(), e);
+            }
+        }
+
+        // Replace path separator characters to avoid inconsistent paths
+        return displayName != null ? displayName.replaceAll("/", "-") : null;
+    }
+
+    private String getDisplayNameFromContentResolver(Uri uri) {
+        String displayName = null;
+        String mimeType = getContentResolver().getType(uri);
+        if (mimeType != null) {
+            String displayNameColumn;
+            if (mimeType.startsWith("image/")) {
+                displayNameColumn = MediaStore.Images.ImageColumns.DISPLAY_NAME;
+            } else if (mimeType.startsWith("video/")) {
+                displayNameColumn = MediaStore.Video.VideoColumns.DISPLAY_NAME;
+
+            } else if (mimeType.startsWith("audio/")) {
+                displayNameColumn = MediaStore.Audio.AudioColumns.DISPLAY_NAME;
+
+            } else {
+                displayNameColumn = MediaStore.Files.FileColumns.DISPLAY_NAME;
+            }
+
+            Cursor cursor = null;
+            try {
+                cursor = getContentResolver().query(
+                        uri,
+                        new String[]{displayNameColumn},
+                        null,
+                        null,
+                        null
+                );
+                if (cursor != null) {
+                    cursor.moveToFirst();
+                    displayName = cursor.getString(cursor.getColumnIndex(displayNameColumn));
+                }
+            } catch (Exception e) {
+                Log.e(TAG, "Could not retrieve display name for " + uri.toString(), e);
+                // nothing else, displayName keeps null
+            } finally {
+                if (cursor != null) {
+                    cursor.close();
+                }
+            }
+        }
+        return displayName;
+    }
```

Comment:
Again, copy-paste. I thought they are self-explanatory: 
- `getDisplayNameFromContentResolver` - gets file name from content resolver. Uses special calls to it to get nice assigned name.
- `getDisplayNameForUri` - get display name for uri (bith `file://` and `content://`).
- `generateDisplayName` - generate new display name. Used if we can't determine shared file name. Can happen for some apps, which don't provide such data.